source/language/gates.rst

```diff
@@ -283,42 +285,93 @@ A gate modifier is a keyword that applies to a gate. A modifier
+controlled value of 0 rather than 1. Mathematically, the negative controlled-:math:`U` gate is
+given by :math:`N_U = I \otimes U^{1-c}`, where :math:`c` is the integer value of the control bit
+and :math:`N_U` is the negative controlled-:math:`U` gate.
 
 .. code-block:: c
 
-   // define x as sqrt(x)
-   gate x q1 {
-       pow(2) @ sx q1;
+   // Define a negative controlled X operation using the negctrl gate modifier.
+   // q1 is control, q2 is target
+   gate neg_cx(θ) q1, q2 {
+       negctrl @ x q1, q2;
    }
 
-The modifier ``ctrl @`` replaces its gate argument :math:`U` by a
-controlled-:math:`U` gate. The new control qubit is prepended to the
-argument list for the controlled-:math:`U` gate. The modified gate does
-not use any additional scratch space. A target may or may not be able to
-execute the gate without further compilation.
+``ctrl`` and ``negctrl`` both accept an optional positive integer argument ``n``, specifying the
+number of control bits (omission means ``n=1``). ``n`` must be a compile-time constant. For an ``N``
+qubit operation, these operations are mathematically defined as
+
+.. math::
+
+   C^n_U = I_1 \otimes I_2 ... \otimes I_n \otimes U^{c_1*c_2*...*c_n}
+
+   N^n_U = I_1 \otimes I_2 ... \otimes I_n \otimes U^{1 - c_1*c_2*...*c_n}
+
+where :math:`c_1`, :math:`c_2`, ..., :math:`c_n` are the integer values of the control bits and
+:math:`C^n_U` are the n-bit controlled-:math:`U` and n-bit negative controlled-:math:`U` gates,
+respectively.
 
 .. code-block:: c
 
-   // Define a controlled Rz operation using the ctrl gate modifier.
-   gate crz(θ) q1, q2 {
-       ctrl @ U(θ, 0, 0) q1, q2;
+   // A reversible boolean function
+   // Demonstrates use of ``ctrl(n) @`` and ``negctrl(n) @``
+   qubit[5] a;
+   qubit f;
+   reset f;
+   negctrl(2) @ ctrl(3) @ x a, f;
+   negctrl(2) @ ctrl(2) @ x a[0], a[3], a[1], a[2], f;
+   negctrl @ ctrl(3) @ x a[0], a[1], a[3], a[4], f;
+   negctrl @ ctrl(3) @ x a[1], a[0], a[3], a[4], f;
+   ctrl(3) @ x a[0], a[1], a[2], f;
+   negctrl(3) @ ctrl @ x a[0], a[1], a[2], a[3], f;
+
+The modifier ``inv @`` replaces its gate argument :math:`U` with its inverse
+:math:`U^\dagger`. This can be computed from gate :math:`U` via the following rules
+
+- The inverse of any gate :math:`U=U_m U_{m-1} ... U_1` can be defined recursively by reversing the
+order of the gates in its definition and replacing each of those with their inverse
+:math:`U^\dagger = U_1^\dagger U_2^\dagger ... U_m^\dagger`.
+- The inverse of a controlled operation is defined by inverting the control unitary. That is,
+``inv @ ctrl @ U = ctrl @ inv @ U``.
+- The base case is given by replacing ``inv @ U(θ, ϕ, λ)`` by ``U(-θ, -λ, -ϕ)`` and
+``inv @ gphase(a)`` by ``gphase(-a)``.
+
+.. code-block:: c
+
+   // Define a negative z rotation and the inverse of a positive z rotation
+   gate rzm(θ) q1 {
+       inv @ rzp(θ) q1;
+   }
+
+The modifier ``pow(k) @`` replaces its gate argument :math:`U` by its :math:`k`\ th
```

Comment:
Should the mathematical definition of power in terms of principal logarithm be added here? (See pp.16 of https://arxiv.org/pdf/2104.14722.pdf) It might be more intuitive to define the `kth` power by raising each eigenvalue to the `kth` power. That is, is we diagonalize the unitary as `U=SDS*` (for unitary `S` and diagonal matrix `D`), then `U^k = SD^kS*`

Reply:
@ajavadia or @awcross1 have an opinion here?

Reply:
The two definitions are essentially the same except that the principal logarithm restricts the arguments of the elements of D to be in (-pi,pi] so that the logarithm is unique. @ajavadia

Reply:
Looks like we can go either way. I am fine to merge as is, but if you prefer to add this, let me know and I will hold off on clicking 'merge'.

Reply:
I am also fine as is, but if we want it changed I'm happy to do so

Reply:
I'm fine with it as is except we might need to say "... for some integer or floating point ..." since it doesn't need to be positive.